source/extensions/filters/listener/tls_inspector/tls_inspector.cc

```diff
@@ -0,0 +1,184 @@
+                                      [this](uint32_t events) {
+                                        ASSERT(events == Event::FileReadyType::Read);
+                                        onRead();
+                                      },
+                                      Event::FileTriggerType::Edge, Event::FileReadyType::Read);
+
+  // TODO(PiotrSikora): make this configurable.
+  timer_ = cb.dispatcher().createTimer([this]() -> void { onTimeout(); });
+  timer_->enableTimer(std::chrono::milliseconds(15000));
+
+  cb_ = &cb;
+  return Network::FilterStatus::StopIteration;
+}
+
+void Filter::onServername(absl::string_view name) {
+  if (!name.empty()) {
+    config_->stats().sni_found_.inc();
+    cb_->socket().setRequestedServerName(name);
+  } else {
+    config_->stats().no_sni_found_.inc();
+  }
+  servername_set_ = true;
+}
+
+void Filter::onRead() {
+  Network::ConnectionSocket& socket = cb_->socket();
+
+  auto& os_syscalls = Api::OsSysCallsSingleton::get();
+  ssize_t n = os_syscalls.recv(socket.fd(), buf_.data(), buf_.capacity(), MSG_PEEK);
+  ENVOY_LOG(trace, "tls inspector: recv: {}", n);
+
+  if (n == -1 && errno == EAGAIN) {
+    return;
+  } else if (n == 0) {
+    // Note: this is not a reliable way to detect EOF. This case can only be
+    // hit if the client closes the connection before writing any bytes, due to
+    // all reads using MSG_PEEK.
+    config_->stats().connection_closed_.inc();
+    done(false);
+    return;
+  } else if (n < 0) {
+    config_->stats().read_error_.inc();
+    done(false);
+    return;
+  }
+
+  // Because we're doing a MSG_PEEK, data we've seen before gets returned everytime, so
+  // skip over what we've already processed.
+  if (static_cast<uint64_t>(n) > read_) {
+    uint8_t* data = buf_.data() + read_;
+    size_t len = n - read_;
+    read_ = n;
+    parseClientHello(data, len);
+  }
+}
+
+void Filter::onTimeout() {
+  ENVOY_LOG(trace, "tls inspector: timeout");
+  config_->stats().read_timeout_.inc();
+  done(false);
+}
+
+void Filter::done(bool success) {
+  ENVOY_LOG(trace, "tls inspector: done: {}", success);
+  timer_.reset();
+  file_event_.reset();
+  cb_->continueFilterChain(success);
+}
+
+void Filter::parseClientHello(void* data, size_t len) {
+  // Ownership is passed to ssl_ in SSL_set_bio()
+  BIO* bio = BIO_new_mem_buf(data, len);
+  SSL_set_bio(ssl_.get(), bio, bio);
+
+  // Make the mem-BIO return that there is more data
+  // available beyond it's end
+  BIO_set_mem_eof_return(bio, -1);
+
+  int ret = SSL_do_handshake(ssl_.get());
+
+  switch (SSL_get_error(ssl_.get(), ret)) {
+  case SSL_ERROR_WANT_READ:
+    if (read_ == buf_.capacity()) {
+      // We've hit the specified size limit. This is an unreasonably large ClientHello;
+      // indicate failure.
+      config_->stats().client_hello_too_big_.inc();
+      done(false);
+    }
+    break;
+  case SSL_ERROR_SSL:
+    if (servername_set_) {
+      done(true);
```

Comment:
Also, I'd still like to see the TLS vs no-TLS being set on the `ConnectionSocket`, like in the original PR.

Reply:
The SNI callback is called even if the SNI extension is not present, but it gives a nullptr for the name in the case where the SNI extension isn't present.

 However, I'd say "servername_set_" is a bad name for this variable. Really, this should be named "clienthello_successfully_processed_" or similar.

Reply:
As for setting TLS vs no-TLS, I think that can be set in onServername(). I'd like to wait until we have the configuration for how to match on that. Otherwise, I think it's likely that whatever I do here won't be quite right. 

I'm adding a TODO in the spot where I think the call should be made.